src/couch_mrview/src/couch_mrview_http.erl

```diff
@@ -709,3 +750,169 @@ maybe_set_page_size(#mrargs{page_size = undefined} = Args, MaxPageSize) ->
     Args#mrargs{page_size = MaxPageSize};
 maybe_set_page_size(#mrargs{} = Args, _MaxPageSize) ->
     Args.
+
+
+bookmark_encode(Args0) ->
+    {Args, _} = lists:foldl(fun(Value, {Acc, Idx}) ->
+        case element(Idx, Acc) of
+            Value ->
+                {setelement(Idx, Acc, []), Idx + 1};
+            _Default when Idx == #mrargs.bookmark
+                -> {setelement(Idx, Acc, []), Idx + 1};
+            _Default ->
+                {setelement(Idx, Acc, Value), Idx + 1}
+        end
+    end, {#mrargs{}, 1}, tuple_to_list(Args0)),
+    Term = erlang:delete_element(1, Args),
+    Binary = term_to_binary(Term, [compressed, {minor_version, 2}]),
+    couch_util:encodeBase64Url(Binary).
+
+
+bookmark_decode(Bookmark) ->
+    Binary = couch_util:decodeBase64Url(Bookmark),
+    Term = binary_to_term(Binary, [safe]),
+    {Args, _} = lists:foldl(fun(Value, {Acc, Idx}) ->
+        case {Value, element(Idx, Acc)} of
+            {[], []} ->
+                {Acc, Idx + 1};
+            {[], Default} ->
+                {setelement(Idx, Acc, Default), Idx + 1};
+            {_, _} when Idx == #mrargs.bookmark ->
+                {setelement(Idx, Acc, true), Idx + 1};
+            {Value, _} ->
+                {setelement(Idx, Acc, Value), Idx + 1}
+        end
+    end, {#mrargs{}, 1}, [mrargs | tuple_to_list(Term)]),
+    Args.
```

Comment:
First things first, this definitely needs to have a versioned field. Even if we don't offer upgrade code (which is fine cause I think we'd all agree that storing bookmarks is not an intended use case) we'll still want to reject with a smarter error message rather than breaking weirdly or worse, responding with data that's misinterpreted.

I'm also not sold on the tuple-to-record implicit conversion. Given that an `mrargs` record is 24 fields, I'd think that most cases would likely be a bit smaller if we encoded the index for non-default values. There's some examples of this sort of approach in `fabric.erl` in the `kl_to_record` function. Though instead of atoms I'd spit out the actual field index so there's no lookup on decode.

Reply:
> First things first, this definitely needs to have a versioned field. 

Agree

> I'm also not sold on the tuple-to-record implicit conversion. Given that an mrargs record is 24 fields, I'd think that most cases would likely be a bit smaller if we encoded the index for non-default values.

I am explicitly converting default values to `[]` which consumes 1 byte per field (see http://erlang.org/doc/apps/erts/erl_ext_dist.html#nil_ext). This is the smallest possible representation I am aware of. The `kl_to_record` is worse for number of fields greater than 3.

```
18> Nil = term_to_binary({1,2,3,4,[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}, [compressed, {minor_version, 2}]).
<<131,80,0,0,0,30,120,156,203,144,72,100,76,100,74,100,78,
  100,201,194,2,0,140,196,10,87>>
19> size(Nil).
26
20> TList = term_to_binary([{1, 1},{2,2},{3,3},{4,4}], [compressed, {minor_version, 2}]).
<<131,108,0,0,0,4,104,2,97,1,97,1,104,2,97,2,97,2,104,2,
  97,3,97,3,104,2,97,4,97,...>>
21> size(TList).
31
```

Reply:
The only smaller encoding I can think of is a 24bit bit mask (3 bytes) in front of a binary which would have `1` for non default fields. Something like:
```
<<Mask:24, (term_to_binary([1,2,3]))/binary>>
```

Reply:
```
23> Fields = term_to_binary([1,2,3], [compressed, {minor_version, 2}]).
<<131,107,0,3,1,2,3>>
24> <<12:24, Fields/binary>>.
<<0,0,12,131,107,0,3,1,2,3>>
```

Reply:
The difference was in eliding the value if it was set to the default so you'd have an extra byte per field, but my assumption was that in most cases a user would be using fewer than half the fields available. Ah, and not wrapping in tuples as well. So it'd be:

```erlang
6> Bin = term_to_binary([1, 1, 2, 2, 3, 3, 4, 4], [compressed, {minor_version, 2}]).
<<131,107,0,8,1,1,2,2,3,3,4,4>>
7> size(Bin).
12
```

However, I think your mask idea is even better. I can see a decode algorithm pretty clearly by converting the mask to a list of integers and then zipping against the list of values. The opposite for encoding seems like it'd be a straightforward reversal of that.